internal/controller/dnsrecord_groups.go

```diff
@@ -0,0 +1,479 @@
+func (s *GroupAdapter) IsActive() bool {
+	//this controller is ungrouped - always active
+	if s.GetGroup() == "" {
+		return true
+	}
+	//this controllers group is active
+	return s.activeGroups.HasGroup(s.GetGroup())
+}
+
+// SetStatusConditions updates the DNSRecord status conditions including the Active condition.
+// The Active condition indicates whether this record's group is currently active:
+//   - ConditionTrue: Record is in an active group (or ungrouped) and will be published
+//   - ConditionFalse: Record is in an inactive group and will not be published
+//   - Condition removed: Record is ungrouped (always active, no condition needed)
+func (s *GroupAdapter) SetStatusConditions(hadChanges bool) {
+	s.DNSRecordAccessor.SetStatusConditions(hadChanges)
+
+	if s.GetGroup() != "" {
+		if s.IsActive() {
+			s.SetStatusCondition(string(v1alpha1.ConditionTypeActive), metav1.ConditionTrue, string(v1alpha1.ConditionReasonInActiveGroup), "Group is included in active groups")
+		} else {
+			s.SetStatusCondition(string(v1alpha1.ConditionTypeActive), metav1.ConditionFalse, string(v1alpha1.ConditionReasonNotInActiveGroup), "Group is not included in active groups")
+		}
+	} else {
+		s.ClearStatusCondition(string(v1alpha1.ConditionTypeActive))
+	}
+}
+
+// getActiveGroups queries DNS for the current list of active groups.
+//
+// It looks up a special TXT record at: kuadrant-active-groups.<zone-domain-name>
+// The TXT record contains semicolon-separated key=value pairs, where the "groups"
+// key contains double-ampersand separated group names.
+//
+// Example TXT record format: "groups=us-east&&us-west;version=1"
+//
+// This function:
+//  1. Constructs the active groups hostname from the zone domain
+//  2. Retrieves custom nameservers from the provider secret (if configured)
+//  3. Queries those nameservers (or defaults) for the TXT record
+//  4. Parses the "groups=..." entry and returns the list of active groups
+//
+// Returns an empty Groups list if:
+//   - The TXT record doesn't exist
+//   - The TXT record is malformed
+//   - DNS lookup fails
+//   - No "groups" key is found in the record
+func (r *BaseDNSRecordReconciler) getActiveGroups(ctx context.Context, c client.Client, dnsRecord DNSRecordAccessor) types.Groups {
+	logger := log.FromContext(ctx).WithName("active-groups")
+	activeGroups := types.Groups{}
+	activeGroupsHost := activeGroupsTXTRecordName + "." + dnsRecord.GetZoneDomainName()
+
+	nameservers, err := r.getNameserversFromProvider(ctx, c, dnsRecord)
+	if err != nil {
+		logger.Error(err, "error getting custom nameservers from provider")
+		return activeGroups
+	}
+	values, err := r.TXTResolver.LookupTXT(ctx, activeGroupsHost, nameservers)
+	if err != nil {
+		logger.Error(err, "error looking up active groups")
+		return activeGroups
+	}
+	// Parse the TXT record to extract active groups
+	// Expected format: "groups=group1&&group2&&group3;version=1;other=value"
```

Comment:
I expected most parts of this format to reside in [constants](https://github.com/philbrookes/dns-operator/blob/gh-620/cmd/plugin/failover/shared.go#L20-L24) and be reused in places we want to reuse them.